pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go

```diff
@@ -122,63 +121,37 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
-	// deployments to allow them to be superceded by the new config version.
-	awaitingCancellations := false
-	if !latestIsDeployed {
-		for i := range existingDeployments {
-			deployment := existingDeployments[i]
-			// Skip deployments with an outcome.
-			if deployutil.IsTerminatedDeployment(deployment) {
-				continue
-			}
-			// Cancel running deployments.
-			awaitingCancellations = true
-			if deployutil.IsDeploymentCancelled(deployment) {
-				continue
-			}
 
-			// Retry faster on conflicts
-			var updatedDeployment *v1.ReplicationController
-			if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
-				rc, err := c.rcLister.ReplicationControllers(deployment.Namespace).Get(deployment.Name)
-				if kapierrors.IsNotFound(err) {
-					return nil
-				}
-				if err != nil {
-					return err
-				}
-				// We need to make sure we own that RC or adopt it if possible
-				isOurs, err := cm.ClaimReplicationController(rc)
-				if err != nil {
-					return fmt.Errorf("error while deploymentConfigController claiming the replication controller %s/%s: %v", rc.Namespace, rc.Name, err)
-				}
-				if !isOurs {
-					return nil
-				}
-
-				copied, err := deployutil.DeploymentDeepCopyV1(rc)
-				if err != nil {
-					return err
-				}
-				copied.Annotations[deployapi.DeploymentCancelledAnnotation] = deployapi.DeploymentCancelledAnnotationValue
-				copied.Annotations[deployapi.DeploymentStatusReasonAnnotation] = deployapi.DeploymentCancelledNewerDeploymentExists
-				updatedDeployment, err = c.rn.ReplicationControllers(copied.Namespace).Update(copied)
-				return err
-			}); err != nil {
-				c.recorder.Eventf(config, v1.EventTypeWarning, "DeploymentCancellationFailed", "Failed to cancel deployment %q superceded by version %d: %s", deployment.Name, config.Status.LatestVersion, err)
-			} else if updatedDeployment != nil {
-				// replace the current deployment with the updated copy so that a future update has a chance at working
-				existingDeployments[i] = updatedDeployment
-				c.recorder.Eventf(config, v1.EventTypeNormal, "DeploymentCancelled", "Cancelled deployment %q superceded by version %d", deployment.Name, config.Status.LatestVersion)
-			}
+	if !latestIsDeployed {
+		if err := c.cancelRunningRollouts(config, existingDeployments, cm); err != nil {
+			return err
 		}
 	}
-	// Wait for deployment cancellations before reconciling or creating a new
-	// deployment to avoid competing with existing deployment processes.
-	if awaitingCancellations {
-		c.recorder.Eventf(config, v1.EventTypeNormal, "DeploymentAwaitingCancellation", "Deployment of version %d awaiting cancellation of older running deployments", config.Status.LatestVersion)
-		return fmt.Errorf("found previous inflight deployment for %s - requeuing", deployutil.LabelForDeploymentConfig(config))
+	// Process triggers and start an initial rollouts
+	configCopy, err := deployutil.DeploymentConfigDeepCopy(config)
```

Comment:
maybe we could do this deep copy at the start of this function and be sure we don't modify shared cache since now the only paths that wouldn't reach it are exceptional